src/main/java/com/snowflake/kafka/connector/internal/SnowflakeSinkServiceV1.java

```diff
@@ -130,6 +135,16 @@ public void startTask(final String tableName, final TopicPartition topicPartitio
 
   @Override
   public void insert(final Collection<SinkRecord> records) {
+    Exception potentialEx = failedCleanerException.get();
+    if (potentialEx != null) {
+      for (ServiceContext pipe : pipes.values()) {
+        // flush current buffers
+        pipe.flushBuffer();
```

Comment:
Should we flush? Flushing here will only increase number of files present on internal stage. the `init()` method also does a list operation on fresh start of the task. 

The committing of offsets happen in `getOffset` when we call ingestFiles API in snowpipe. 

Also, if we are sending connect exception, all the `ServiceContext` objects will be reinitialized because we will have to restart the task.. Restart task-> `start()` method of SinkTask?

Reply:
> Also, if we are sending connect exception, all the ServiceContext objects will be reinitialized because we will have to restart the task.. Restart task-> start() method of SinkTask? (Correct me if this is not the right API called on task restart) 

Yes this is right. Once the task fails - and if it's restarted by the user- we would reinit the `ServiceContext` objects. 

So the initial reasoning behind this was I thought we were committing offsets for successful write to the stage. But after another pass, I think we're committing offsets only based on successful ingestion into snowpipe. Therefore we can rely on the Kafka Connect framework to provide us with the right records after a restart so that we can avoid data loss. 

In fact, flushing the buffer is incorrect since it might end up writing duplicates to the stage - which might betray EOS semantics. (though I guess with a consistent naming scheme for the files in the stage - this write could be idempotent).  

I'll get rid of the flush!

Reply:
Sounds good!

Reply:
It's actually not idempotent since the file name has timestamp of flush. So it is probably better to not flush (Thanks for changing it)

Reply:
> So the initial reasoning behind this was I thought we were committing offsets for successful write to the stage. But after another pass, I think we're committing offsets only based on successful ingestion into snowpipe. 

Are you sure this is correct? We observed that commiting only after confirmation of the ingestion-report that snowpipe has processed the files and their contents have made it to the snowflake table was _pre v1.4 behaviour_. Since this PR (https://github.com/snowflakedb/snowflake-kafka-connector/pull/152), it seemed to us that the connector intentionally changed to a more aggressive (and potentially premature) offset strategy where it commits immediately after records are flushed (and an insert-request is made to snowpipe).

I think this implies your initial observation is correct - if snowpipe fails to ingest records on the stage and it kills the connector tasks, when the user restarts the task, Kafka will not resend those failed records. They will be left in  orphaned files on the stage. They likely won't be purged automatically by `filterFileReprocess`, but they also won't be reingested by snowpipe since there is nothing to make an ingest-request on its behalf. This seems pretty alarming to me as it suggests that there is _effectively_ data-loss here. The connectors job is to get data from Kafka to snowflake tables, not merely to snowflake _stages_. It seems dangerous then to consider killing the connector task without moving the current offset-commit strategy back to the more conservative approach pre v1.4, no?